text/0000-macros-1.2.md

```diff
@@ -0,0 +1,167 @@
+- Feature Name: macros-1.2
+- Start Date: 2017-02-20
+- RFC PR:
+- Rust Issue:
+
+# Summary
+[summary]: #summary
+
+Stabilize function-like procedural macros (whose usage looks like `foo!(...)`),
+like this was done in “[Macros 1.1]” for custom `derive`,
+before “[Macros 2.0]” is fully ready.
+
+[Macros 1.1]: https://github.com/rust-lang/rfcs/blob/master/text/1681-macros-1.1.md
+[Macros 2.0]: https://github.com/rust-lang/rfcs/blob/master/text/1566-proc-macros.md
+
+
+# Motivation
+[motivation]: #motivation
+
+The full design of Macros 2.0 has many details (around hygiene, the `TokenStream` API, etc.)
+that will require a significant amount of work before it can be fully stabilized.
+
+With Macros 1.1, we chose to stabilize a very small part of the new API
+that was nevertheless enough to unlock a significant portion of the benefits.
+This RFC propose what is comparatively a small additional step,
+while also enabling new use cases.
+
+At the moment, like they used to for custom derive, some crates resort to [complicated schemes]
+that involve parsing entire source files with the `syn` crate,
+manually expanding a macro, and using the generated file through `include!()`.
+This approach is viable (if inconvenient) within one crate for one source file,
+but is probably not acceptable for having a library provide a procedural macro
+to be used in other projects.
+
+With this RFC accepted and implemented,
+libraries running on Rust’s stable channel would be able to export procedural macros
+that are as convenient to use as custom derive is since Rust 1.15.
+
+While the use cases for this may not be as prevalent or high-profile as Serde or Diesel,
+the additional amount of details being stabilized
+(compared to what is already stable with Macros 1.1)
+is also very small.
+
+[complicated schemes]: https://github.com/servo/html5ever/blob/e29d495c94/macros/match_token.rs
+
+
+# Detailed design
+[design]: #detailed-design
+
+As a reminder, Macro 1.1 stabilized a new `proc_macros` crate with a very small public API:
+
+```rust
+pub struct TokenStream { /* private */ }
+impl fmt::Display for TokenStream {}
+impl FromStr for TokenStream {
+    type Err = LexError;
+}
+pub struct LexError { /* private */ }
+```
+
+As well as an attribute for defining custom derives:
+
+```rust
+#[proc_macro_derive(Example)]
+pub fn example(input: TokenStream) -> TokenStream {
+    // ...
+}
+```
+
+Until more APIs are stabilized for `TokenStream`,
+procedural macros are expected to serialize it to a string
+and parse the result, for example with the [syn](https://github.com/dtolnay/syn) crate.
+
+This RFC does *not* propose any such API.
+It propose prioritizing the implementation and stabilization
+of function-like procedural macros, that are defined like this:
+
+```rust
+#[proc_macro]
+pub fn foo(input: TokenStream) -> TokenStream {
+    // ...
+}
+```
+
+And used (in a separate crate that depends on the previous one) like this:
+
+```rust
+foo!(...);
```

Comment:
Are there any constraints as to what these can expand to? Items/statements/expressions/patterns/etc?